pkg/osquery/tables/xfconf/parser.go

```diff
@@ -0,0 +1,100 @@
+//go:build !windows
+// +build !windows
+
+package xfconf
+
+import (
+	"encoding/xml"
+	"fmt"
+	"os"
+
+	"github.com/kolide/launcher/pkg/dataflatten"
+)
+
+type (
+	ArrayValue struct {
+		Type  string `xml:"type,attr"`
+		Value string `xml:"value,attr"`
+	}
+	Property struct {
+		Name       string       `xml:"name,attr"`
+		Type       string       `xml:"type,attr"`
+		Value      string       `xml:"value,attr"`
+		Properties []Property   `xml:"property"`
+		Values     []ArrayValue `xml:"value"`
+	}
+	ChannelXML struct {
+		XMLName     xml.Name   `xml:"channel"`
+		ChannelName string     `xml:"name,attr"`
+		Properties  []Property `xml:"property"`
+	}
+)
+
+// parseXml reads in the given xml file, parses it as an xfconf XML file, and then flattens
+// it. Because most XML elements in the xfconf files are `property`, we parse the file into
+// a map with the name attributes set as the keys to avoid loss of meaningful full keys.
+func parseXfconfXml(file string, opts ...dataflatten.FlattenOpts) ([]dataflatten.Row, error) {
+	channelXml, err := readChannelXml(file)
+	if err != nil {
+		return nil, fmt.Errorf("could not read xfconf channel file %s: %w", file, err)
+	}
+
+	return dataflatten.Flatten(channelXml.toMap(), opts...)
+}
+
+func readChannelXml(file string) (ChannelXML, error) {
+	rdr, err := os.Open(file)
+	if err != nil {
+		return ChannelXML{}, err
+	}
+
+	xmlDecoder := xml.NewDecoder(rdr)
+
+	var result ChannelXML
+	xmlDecoder.Decode(&result)
+
+	return result, nil
+}
+
+// toMap transforms Result r into a map where the top-level key is "channel/<name>".
+func (c ChannelXML) toMap() map[string]interface{} {
```

Comment:
This seems fine, though I'm somewhat surprised we need it. You're starting with a well formed `ChannelXML` object? We _ought_ be able to convert it. On the other hand, I bet this works. So 🤷

Reply:
I didn't try! Let me see how that works

Reply:
Seeing this:

```
flattening at path []: unknown type on [{{ channel} thunar-volman [{automount-media empty  [{enabled bool false [] []}] []} {automount-drives empty  [{enabled bool false [] []}] []} {autobrowse empty  [{enabled bool false [] []}] []} {autoopen empty  [{enabled bool false [] []}] []}]} {{ channel} xfce4-power-manager [{xfce4-power-manager empty  [{power-button-action empty  [] []} {lock-screen-suspend-hibernate bool true [] []}] []}]}]
```

Reply:
Yeah... There's something weird in the `Flatten` routine. You can marshal to json, and use FlattenJson, or leave the `toMap` stuff. I suspect we should think a bit harder someday

Reply:
I think I'd like to keep `toMap`, then, since it gives us the nice `channel/<name>` top-level key

Reply:
👍